### PR TITLE
feat: move controller informer-related configuration to InformerConfig

### PR DESCRIPTION
**Pull request by metacosm**

### CODE REVIEW

**Review by csviri — CHANGES_REQUESTED** [review]
Made some changes, I think we should discuss mainly some naming and API issues mentioned, mainly the one related to the `followControllerNamespacesOnChange`.

Shall we just document that, that it is a problem. Maybe we can check and throw an exception if that is set explicitly for the controller?!

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/InformerConfig.java [hunk 1]
```diff
@@ -21,56 +24,60 @@
   String name() default NO_VALUE_SET;
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/InformerConfig.java:24]: This does not makes too much sense for the controller either :(

File: sample-operators/webpage/src/main/java/io/javaoperatorsdk/operator/sample/WebPageReconciler.java [hunk 2]
```diff
@@ -42,19 +42,19 @@ public WebPageReconciler() {
...
-            .withLabelSelector(SELECTOR)
+            .withInformerConfiguration(c -> c.withLabelSelector(SELECTOR))
             .build(), context);
     var serviceEventSource =
         new InformerEventSource<>(InformerConfiguration.from(Service.class, WebPage.class)
-            .withLabelSelector(SELECTOR)
+            .withInformerConfiguration(c -> c.withLabelSelector(SELECTOR))
             .build(), context);
     var ingressEventSource =
         new InformerEventSource<>(InformerConfiguration.from(Ingress.class, WebPage.class)
```
[the remark this answers is not in the document]
Comment (metacosm) [reply sample-operators/webpage/src/main/java/io/javaoperatorsdk/operator/sample/WebPageReconciler.java:56]: The idea is to reuse the same code at some point to deal with the InformerConfig… and yes, the naming is sub-optimal, which is why I'd like to clean things up but I'd rather do that in a subsequent PR

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/informer/InformerConfiguration.java [hunk 3]
```diff
@@ -27,42 +25,79 @@ public interface InformerConfiguration<R extends HasMetadata>
...
+
+  /**
+   * Used in case the watched namespaces are changed dynamically, thus when operator is running (See
+   * {@link io.javaoperatorsdk.operator.RegisteredController}). If true, changing the target
+   * namespaces of a controller would result to change target namespaces for the
+   * InformerEventSource.
+   *
+   * @return if namespace changes should be followed
+   */
+  default boolean followControllerNamespaceChanges() {
```
[the remark this answers is not in the document]
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/informer/InformerConfiguration.java:50]: I know but I think this should be dealt with in a subsequent PR.

File: sample-operators/webpage/src/main/java/io/javaoperatorsdk/operator/sample/WebPageStandaloneDependentsReconciler.java [hunk 4]
```diff
@@ -96,8 +96,9 @@ private Workflow<WebPage> createDependentResourcesAndWorkflow() {
     // configure them with our label selector
     Arrays.asList(configMapDR, deploymentDR, serviceDR, ingressDR)
         .forEach(dr -> dr.configureWith(new KubernetesDependentResourceConfigBuilder()
-            .withKubernetesDependentInformerConfig(new KubernetesDependentInformerConfigBuilder<>()
-                .withLabelSelector(SELECTOR + "=true").build())
+            .withKubernetesDependentInformerConfig(InformerConfigHolder.builder()
```
[the remark this answers is not in the document]
Comment (metacosm) [reply sample-operators/webpage/src/main/java/io/javaoperatorsdk/operator/sample/WebPageStandaloneDependentsReconciler.java:99]: Agreed but I'd rather rename things in a separate cleaning up phase, if we agree with the overall change.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/InformerConfig.java (hunk 1 above)
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/InformerConfig.java:24]: I wondered about this, actually, if we could somehow reuse the same config for the controller and some other event source but yes, I agree that this doesn't make much sense for the controller.

The problem is that we cannot inherit between annotations so either:  

- we have separate annotations that do more or less the same thing with code to process each that will again more or less do the same thing, which will cause quite a big maintenance load
- we accept that some of the setup doesn't make sense for both, add proper safeguards and live with it
-  we extract the fields that don't make sense in a different annotation for event sources that will enclose an `InformerConfig` field but that adds yet another level of annotations, which would be less than ideal.

**Comment by metacosm** [comment]
> Made some changes, I think we should discuss mainly some naming and API issues mentioned, mainly the one related to the `followControllerNamespacesOnChange`.

I don't see any changes. Did you push them?

> 
> Shall we just document that, that it is a problem. Maybe we can check and throw an exception if that is set explicitly for the controller?!

Maybe.

File: sample-operators/webpage/src/main/java/io/javaoperatorsdk/operator/sample/WebPageReconciler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (csviri) [reply sample-operators/webpage/src/main/java/io/javaoperatorsdk/operator/sample/WebPageReconciler.java:56]: I mean, we could just let the old methods, like here, the `withLabelSelector` will be in the background the `InformerConfigHolder`. Would be good to have this in this PR, this is something I would not release for sure. But if you insist feel free to do separately.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/InformerConfig.java (hunk 1 above)
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/InformerConfig.java:24]: Exactly, imo 1. or 2.
Not sure if that is that much more code to have a `@ControllerInformer`. 

Or alternative is also just simply stick with the current approach and don't merge this PR.

File: sample-operators/webpage/src/main/java/io/javaoperatorsdk/operator/sample/WebPageReconciler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply sample-operators/webpage/src/main/java/io/javaoperatorsdk/operator/sample/WebPageReconciler.java:56]: I think I had tried that (or some variation of it) but maybe I didn't. Let me see what it does.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/InformerConfig.java (hunk 1 above)
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/InformerConfig.java:24]: I think I would prefer 2. at this point, with more clean up on the naming, it should be much better, I just wanted to control the blast radius at this point :)

File: sample-operators/webpage/src/main/java/io/javaoperatorsdk/operator/sample/WebPageReconciler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply sample-operators/webpage/src/main/java/io/javaoperatorsdk/operator/sample/WebPageReconciler.java:56]: I would re-add these methods after we're done with renaming and cleaning-up if we feel it's necessary.

**Review by csviri — APPROVED** [review]

**Comment by metacosm** [comment]
Thanks, I will open a subsequent PR to clean things up and rename things more appropriately.

**Pull request merged**